qcodes/dataset/sqlite/queries.py

```diff
@@ -332,49 +335,50 @@ def _expand_data_to_arrays(
-                for element in text_elms:
-                    strlen = len(row[element])
-                    row[element] = np.full_like(row[first_array_element],
-                                                row[element],
-                                                dtype=np.dtype(f'U{strlen}'))
-
-        for row in data:
+            types_mapping: dict[int, Callable[[str], np.dtype[Any]]] = {}
+            for i, x in enumerate(types):
+                if x == "numeric":
+                    types_mapping[i] = lambda _: np.dtype(np.float64)
+                elif x == "complex":
+                    types_mapping[i] = lambda _: np.dtype(np.complex128)
+                elif x == "text":
+                    types_mapping[i] = lambda array: np.dtype(f"U{len(array)}")
+
+            for i_row, row in enumerate(data):
+                # todo should we handle int/float types here
+                # we would in practice have to perform another
+                # loop to check that all elements of a given can be cast to
+                # int without loosing precision before choosing an integer
+                # representation of the array
+                data[i_row] = tuple(
+                    np.full_like(
+                        row[first_array_element], array, dtype=types_mapping[i](array)
+                    )
+                    if i in types_mapping
+                    else array
+                    for i, array in enumerate(row)
+                )
+
+        for i_row, row in enumerate(data):
             # now expand all one element arrays to match the expected size
             # one element arrays are introduced if scalar values are stored
             # with an explicit array storage type
-            sizes = tuple(array.size for array in row)
-            max_size = max(sizes)
-            max_index = sizes.index(max_size)
-
+            max_size = 0
             for i, array in enumerate(row):
-                if array.size != max_size:
-                    if array.size == 1:
-                        row[i] = np.full_like(row[max_index],
-                                              row[i],
-                                              dtype=row[i].dtype)
-                    else:
-                        log.warning(f"Cannot expand array of size {array.size} "
-                                    f"to size {row[max_index].size}")
+                if array.size > max_size:
+                    if max_size > 1:
+                        log.warning(
+                            f"Cannot expand array of size {max_size} "
+                            f"to size {array.size}"
+                        )
+                    max_size, max_index = array.size, i
+
+            data[i_row] = tuple(
+                np.full_like(row[max_index], array, dtype=array.dtype)
+                if array.size != max_size
+                else array
+                for array in row
+            )
```

Comment:
Before:
https://github.com/QCoDeS/Qcodes/blob/01951b5684d50087e895b80e88ce9b60c9779989/qcodes/dataset/sqlite/queries.py#L361-L377

After:
https://github.com/QCoDeS/Qcodes/blob/ec9bd8ca805c58affeca28f32f84e1c5eae2a980/qcodes/dataset/sqlite/queries.py#L362-L382

The algorithm is the same, it has to be rewritten because each row is now a tuple and cannot be modified directly. Note the following:
- Using `full` vs `full_like` is a bit faster
- The warning case is never tested